Microsoft.NET.Build.Containers/Registry.cs

```diff
@@ -82,39 +98,112 @@ public Registry(Uri baseUri)
+        var runtimeGraph = GetRuntimeGraphForDotNet(runtimeIdentifierGraphPath);
+        var (ridDict, graphForManifestList) = ConstructRuntimeGraphForManifestList(manifestList, runtimeGraph);
+        var bestManifestRid = CheckIfRidExistsInGraph(graphForManifestList, runtimeIdentifier);
+        if (bestManifestRid is null) {
+            throw new ArgumentException($"The runtimeIdentifier '{runtimeIdentifier}' is not supported. The supported RuntimeIdentifiers for the base image {repositoryName}:{reference} are {String.Join(",", graphForManifestList.Runtimes.Keys)}");
+        }
+        var matchingManifest = ridDict[bestManifestRid];
+        var manifestResponse = await GetManifest(repositoryName, matchingManifest.digest);
+        return await TryReadSingleImage(repositoryName, await manifestResponse.Content.ReadFromJsonAsync<ManifestV2>());
+    }
 
-        if (manifest is null) throw new NotImplementedException("Got a manifest but it was null");
+    private async Task<HttpResponseMessage> GetManifest(string repositoryName, string reference)
+    {
+        var client = GetClient();
+        var response = await client.GetAsync(new Uri(BaseUri, $"/v2/{repositoryName}/manifests/{reference}"));
+        response.EnsureSuccessStatusCode();
+        return response;
+    }
 
-        if ((string?)manifest["mediaType"] != DockerManifestV2)
-        {
-            throw new NotImplementedException($"Do not understand the mediaType {manifest["mediaType"]}");
-        }
+    async Task<HttpResponseMessage> GetBlob(string repositoryName, string digest)
+    {
+        var client = GetClient();
+        var response = await client.GetAsync(new Uri(BaseUri, $"/v2/{repositoryName}/blobs/{digest}"));
+        response.EnsureSuccessStatusCode();
+        return response;
+    }
 
-        JsonNode? config = manifest["config"];
-        Debug.Assert(config is not null);
-        Debug.Assert(((string?)config["mediaType"]) == DockerContainerV1);
+    private string? CheckIfRidExistsInGraph(RuntimeGraph graphForManifestList, string userRid) => graphForManifestList.Runtimes.FirstOrDefault(kvp => graphForManifestList.AreCompatible(kvp.Key, userRid)).Key;
 
-        string? configSha = (string?)config["digest"];
-        Debug.Assert(configSha is not null);
+    private (IReadOnlyDictionary<string, PlatformSpecificManifest>, RuntimeGraph) ConstructRuntimeGraphForManifestList(ManifestListV2 manifestList, RuntimeGraph dotnetRuntimeGraph)
+    {
+        var ridDict = new Dictionary<string, PlatformSpecificManifest>();
+        var runtimeDescriptionSet = new HashSet<RuntimeDescription>();
+        foreach (var manifest in manifestList.manifests) {
+            if (CreateRidForPlatform(manifest.platform) is { } rid)
+            {
+                if (ridDict.TryAdd(rid, manifest)) {
```

Comment:
Are there cases where there are more than one? Is taking the first one and ignoring subsequent always right? Should we throw instead?

Reply:
There's a tension between the RID graph and the image lists here - for the common case of the windows .NET base images, there are two windows' available. Here's dotnet/runtime:7.0

```json
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1160,
         "digest": "sha256:3bc8d3de62b630cd8e39431b825686558333374f455ab5f804db17ffff893ebc",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1160,
         "digest": "sha256:636a9588ee4dc82bdaf65521adf72d4d9bd5cea723dc8fef43045073dcc25342",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v7"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1160,
         "digest": "sha256:7bf33f12e8a0e4ebc887da9b84607b76f2fd609a23578faf3c491d0a5a66c205",
         "platform": {
            "architecture": "arm64",
            "os": "linux",
            "variant": "v8"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1754,
         "digest": "sha256:7398943dbd3b0ed46afc6aa2fad65d4bd287f74efa3ca55e7b6872cbc59b7c52",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.17763.3887"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1754,
         "digest": "sha256:6f6425cf3da9662483737b46e9992bc7e7c9fbfe6a693b49db57e8c912ae4b1d",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.20348.1487"
         }
      }
   ]
}
```

We end up with two manifests that could be `win-x64` here, and the RID graph doesn't go precise enough to differentiate.  So I settled for the first one.

I think my answer to users who are confused by this will end up being to ask them to specify the base image exactly, since these same images are available via tags: `mcr.microsoft.com/dotnet/runtime:7.0.2-windowsservercore-ltsc2019`.